### ISSUE TITLE
🚀 Feature: Make Catalog side filters customizable to show/hide as needed

### ISSUE THREAD

**Issue by solimant** [backstage/backstage#17172]
### 🔖 Feature description

Would like a way to be able to show/hide certain Catalog side filters as needed:
![image](https://user-images.githubusercontent.com/11140840/228866515-f3ddce85-ca82-47a6-ba5c-3324e2e7a048.png)

### 🎤 Context

In our use-case, certain pickers on the Catalog page are either not relevant, or there's no enough data to keep that filter. For example, the _Kind_ picker may not be necessary if the Catalog contains only one kind. Similarly, the _Processing Status_ picker. The workaround to make this happen today is to spin off a custom Catalog page, which according to the docs, is not really recommended, unless it's a very special case. The show/hide of certain side filters seems trivial enough to have to spin off a custom Catalog page.

### ✌️ Possible Implementation

Extend the `DefaultCatalogPageProps` props to include optional props for which pickers to hide, where the default is to show everything as it does today. For example, in [`DefaultCatalogPage.tsx`](https://github.com/backstage/backstage/blob/7fe57389c6d0eb3b37f5b27c2d429aff09004041/plugins/catalog/src/components/CatalogPage/DefaultCatalogPage.tsx):
```diff
+type FilterKind =
+ | 'kind'
+ | 'lifecycle'
+ | 'owner'
+ | 'processing_status'
+ | 'tag'
+ | 'type'
+ | 'user';

export interface DefaultCatalogPageProps {
  initiallySelectedFilter?: UserListFilterKind;
  columns?: TableColumn<CatalogTableRow>[];
  actions?: TableProps<CatalogTableRow>['actions'];
  initialKind?: string;
  tableOptions?: TableProps<CatalogTableRow>['options'];
  emptyContent?: ReactNode;
+ filtersToHide?: FilterKind[];
}

/* ... further later ... */
<CatalogFilterLayout.Filters>
  <EntityKindPicker
    initialFilter={initialKind}
+   hidden={filtersToHide?.includes('kind')}
/>
  <EntityTypePicker
+   hidden={filtersToHide?.includes('type')}
/>
  {/* ... */}
</CatalogFilterLayout.Filters>
```

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by benjdlambert** [comment]
So I think that we would rather push you to using the existing `CatalogPage` customization that enables you to provide your own `CatalogIndexPage` with whatever filters you want. https://backstage.io/docs/features/software-catalog/catalog-customization/

**Comment by solimant** [comment]
Thanks, Ben. Yes, that’s what we’re doing, but there’s a note on that page
that says:
> Note: The catalog index page is designed to have a minimal code footprint
to support easy customization, but creating a copy does introduce a
possibility of drifting out of date over time. Be sure to check the catalog
CHANGELOG
<https://github.com/backstage/backstage/blob/master/plugins/catalog/CHANGELOG.md>
 periodically.

Doesn’t that put a burden on us to always make sure we’re aligned? With
that in mind, what’s the concern with extending the API, as described in
this feature request?

**Comment by benjdlambert** [comment]
@solimant yeah, it's basically there as a warning that the API for the customization might change over time but it's designed to be pretty flexible so that you can do this sort of things without changing the core offering.

We're looking at revisiting some of the customization and composibility of all plugins in the future, so I'd want to push back on expanding the API surface any more until we have addressed that over in here #16876
